libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/VoiceInstructionsPrefetcher.kt

```diff
@@ -0,0 +1,139 @@
+package com.mapbox.navigation.ui.voice.api
+
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.base.trip.model.RouteProgress
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.directions.session.RoutesExtra
+import com.mapbox.navigation.core.directions.session.RoutesObserver
+import com.mapbox.navigation.core.directions.session.RoutesUpdatedResult
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationApp
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationObserver
+import com.mapbox.navigation.core.trip.session.RouteProgressObserver
+import com.mapbox.navigation.utils.internal.Time
+import com.mapbox.navigation.utils.internal.ifNonNull
+
+/**
+ * Class that is responsible for listening to relevant updates and triggering
+ * voice instructions predownloading when needed.
+ * Register and unregister with [MapboxNavigationApp.registerObserver]
+ * and [MapboxNavigationApp.unregisterObserver]
+ * or invoke [onAttached] and [onDetached] manually if you are not using [MapboxNavigationApp].
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+class VoiceInstructionsPrefetcher internal constructor(
+    private val speechApi: MapboxSpeechApi,
+    private val observableTime: Int,
+    private val timePercentageToTriggerAfter: Double,
+    private val nextVoiceInstructionsProvider: NextVoiceInstructionsProvider =
+        TimeBasedNextVoiceInstructionsProvider(observableTime),
+    private val timeProvider: Time = Time.SystemClockImpl,
+) : MapboxNavigationObserver {
+
+    /**
+     * Creates [VoiceInstructionsPrefetcher.
+     *
+     * @param speechApi [MapboxSpeechApi] instances that's used to generate instructions
+     */
+    constructor(speechApi: MapboxSpeechApi) : this(
+        speechApi,
+        DEFAULT_OBSERVABLE_TIME_SECONDS,
+        DEFAULT_TIME_PERCENTAGE_TO_TRIGGER_AFTER
+    )
+
+    private val routesObserver = RoutesObserver { onRoutesChanged(it) }
+
+    private val routeProgressObserver = RouteProgressObserver { onRouteProgressChanged(it) }
+
+    private val ignoredRouteUpdateReasons = setOf(
+        RoutesExtra.ROUTES_UPDATE_REASON_CLEAN_UP,
+        RoutesExtra.ROUTES_UPDATE_REASON_ALTERNATIVE,
+        RoutesExtra.ROUTES_UPDATE_REASON_REFRESH,
```

Comment:
Should we trigger predownload immediately after a route refresh? If I understand correctly, voice instructions can change as a result of it.

Reply:
How can they change? I don't see any way.

Reply:
Looks like I was wrong, according to the docs only traffic annotations change after a route refresh.